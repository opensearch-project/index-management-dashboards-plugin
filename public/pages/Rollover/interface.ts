/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { FieldInstance } from "../../lib/field";
import { IndexItem } from "../../models/interfaces";

export interface SubDetailProps {
  field: FieldInstance;
  sourceType?: "dataStreams" | "alias" | undefined;
  writingIndex?: string;
}

export interface IRolloverRequestBody {
  source?: string;
  targetIndex?: IndexItem;
  conditions?: {
    max_age?: string;
    max_docs?: number;
    max_size?: string;
    max_primary_shard_size?: string;
  };
}
