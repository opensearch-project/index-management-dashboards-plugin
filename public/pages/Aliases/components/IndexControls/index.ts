/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexControls from "./IndexControls";

export * from "./IndexControls";
export default IndexControls;
