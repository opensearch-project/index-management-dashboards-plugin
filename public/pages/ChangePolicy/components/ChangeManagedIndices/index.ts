/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ChangeManagedIndices from "./ChangeManagedIndices";

export default ChangeManagedIndices;
