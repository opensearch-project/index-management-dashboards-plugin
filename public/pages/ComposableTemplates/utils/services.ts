/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonService } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";

export const getAllUsedComponents = async ({ commonService }: { commonService: CommonService }) => {
  const allTemplatesResponse = await commonService.apiCaller<{
    index_templates?: Array<{
      name: string;
      index_template: TemplateItemRemote;
    }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: "/_index_template/*",
    },
  });

  if (allTemplatesResponse && allTemplatesResponse.ok) {
    return allTemplatesResponse.response.index_templates || [];
  }

  return [];
};
