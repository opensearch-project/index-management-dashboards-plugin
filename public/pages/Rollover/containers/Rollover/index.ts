/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import IndexDetail from "./Rollover";

export default IndexDetail;
