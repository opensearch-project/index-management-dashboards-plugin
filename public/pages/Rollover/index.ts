/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Rollover from "./containers/Rollover";

export default Rollover;
