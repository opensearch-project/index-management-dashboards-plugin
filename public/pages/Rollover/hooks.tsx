/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { EuiFormRowProps } from "@elastic/eui";
import { merge, isEmpty } from "lodash";
import { FieldInstance, transformNameToString } from "../../lib/field";
import { IFormGeneratorRef } from "../../components/FormGenerator";
import { IndexForm } from "../../containers/IndexForm";
import { IRolloverRequestBody } from "./interface";
import { IndexItemRemote } from "../../../models/interfaces";
import { BrowserServices } from "../../models/interfaces";
import { Alias } from "../../../server/models/interfaces";
import { ServerResponse } from "../../../server/models/types";

export type TypeSourceType = "dataStreams" | "alias" | undefined;

export const getCommonFormRowProps = (name: string | string[], field: FieldInstance): Partial<EuiFormRowProps> => {
  return {
    isInvalid: !!field.getError(name),
    error: field.getError(name),
    "data-test-subj": `form-row-${transformNameToString(name)}`,
  };
};

export const onSubmit = async (props: {
  sourceRef: React.RefObject<IFormGeneratorRef>;
  sourceType: TypeSourceType;
  writingIndex: string;
  indexFormRef: React.RefObject<IndexForm>;
  tempValue: IRolloverRequestBody;
  services: BrowserServices;
}): Promise<
  ServerResponse<{
    new_index: string;
  }>
> => {
  const { sourceRef, sourceType, writingIndex, indexFormRef, tempValue, services } = props;
  const formGeneratersRes = await Promise.all([
    sourceRef.current?.validatePromise(),
    sourceType === "alias" && !writingIndex.match(/^.*-\d+$/)
      ? (new Promise(async (resolve) => {
          const result = await indexFormRef.current?.validate();
          if (result) {
            resolve({ errors: { targetIndex: [result] }, values: {} });
          } else {
            resolve({
              errors: null,
              values: {
                targetIndex: indexFormRef.current?.getValue() || {},
              },
            });
          }
        }) as Promise<{ errors: Record<string, string[]> | null; values: Record<string, any> }>)
      : Promise.resolve({
          errors: null,
          values: {
            targetIndex: indexFormRef.current?.getValue() || {},
          },
        }),
  ]);
  const hasError = formGeneratersRes.some((item) => item?.errors);
  if (hasError) {
    return {
      ok: false,
      error: "Some fields does not pass validation, please check fields with red underline.",
    };
  }

  const finalValues: IRolloverRequestBody = merge({}, tempValue);
  formGeneratersRes.forEach((item) => merge(finalValues, item?.values));

  const payload: {
    alias: string;
    newIndex?: string;
    body: Omit<IndexItemRemote, "index"> & { conditions?: IRolloverRequestBody["conditions"] };
  } = {
    alias: finalValues.source || "",
    body: {},
  };

  if (sourceType === "alias" && !isEmpty(finalValues.targetIndex || {})) {
    const { index, ...others } = (finalValues.targetIndex || {}) as IndexItemRemote;
    if (index) {
      payload.newIndex = index;
    }
    payload.body = others;
  }

  const result = await services.commonService.apiCaller<{
    new_index: string;
  }>({
    endpoint: "indices.rollover",
    data: payload,
  });

  return result;
};

export const submitWriteIndex = async (props: {
  services: BrowserServices;
  writeIndexValue: string;
  sourceRef: React.RefObject<IFormGeneratorRef>;
}) => {
  const { services, writeIndexValue, sourceRef } = props;
  return await services.commonService.apiCaller({
    endpoint: "indices.updateAliases",
    data: {
      body: {
        actions: [
          {
            add: {
              index: writeIndexValue,
              alias: sourceRef.current?.getValue("source"),
              is_write_index: true,
            },
          },
        ],
      },
    },
  });
};

export const getOptions = async (props: { services: BrowserServices }) => {
  const { services } = props;
  return Promise.all([
    services.commonService.apiCaller<Alias[]>({
      endpoint: "cat.aliases",
      data: {
        format: "json",
      },
    }),
    services.indexService.getDataStreams({}),
  ]).then(([aliases, dataStreams]) => {
    if (aliases.ok && dataStreams.ok) {
      const allAlias: { label: string; aliases: Alias[] }[] = [];
      aliases.response.forEach((item) => {
        let findIndex;
        if (allAlias.find((alias) => alias.label === item.alias)) {
          findIndex = allAlias.findIndex((alias) => alias.label === item.alias);
        } else {
          findIndex = allAlias.length;
          allAlias.push({
            label: item.alias,
            aliases: [],
          });
        }
        allAlias[findIndex].aliases.push(item);
      });
      return {
        ok: true,
        response: {
          alias: allAlias,
          dataStreams: dataStreams.response?.dataStreams.map((item) => ({ label: item.name })) || [],
        },
      };
    }

    return {
      ok: false,
      error: aliases.error || dataStreams.error,
    };
  });
};

export const getRolloveredIndex = async (props: { alias: string; services: BrowserServices }): Promise<ServerResponse<string>> => {
  const { alias, services } = props;
  const result = await services.commonService.apiCaller<{
    new_index: string;
  }>({
    endpoint: "indices.rollover",
    data: {
      alias,
      dry_run: true,
    },
    hideLog: true,
  });
  if (result.ok) {
    return {
      ok: true,
      response: result.response.new_index,
    };
  } else if (result && result.body && result.body.error && result.body.error.index) {
    return {
      ok: true,
      response: result.body.error.index,
    };
  }

  return {
    ok: false,
    error: result.error,
  };
};

export const getIndexDetail = async ({
  services,
  indexName,
  newIndexName,
}: {
  services: BrowserServices;
  indexName: string;
  newIndexName: string;
}) => {
  const response = await services.commonService.apiCaller<Record<string, IndexItemRemote>>({
    endpoint: "indices.get",
    data: {
      index: indexName,
      flat_settings: true,
    },
  });
  if (response.ok) {
    return {
      ...response.response[indexName],
      index: newIndexName,
    };
  }

  throw new Error(response.error);
};
