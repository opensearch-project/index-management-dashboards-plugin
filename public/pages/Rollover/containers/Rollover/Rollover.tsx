/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useContext, useEffect, useMemo, useRef, useState } from "react";
import { RouteComponentProps } from "react-router-dom";
import { merge } from "lodash";
import { BREADCRUMBS, ROUTES } from "../../../../utils/constants";
import { ServicesContext } from "../../../../services";
import { BrowserServices } from "../../../../models/interfaces";
import { CoreServicesContext } from "../../../../components/core_services";
import IndexFormWrapper, { IndexForm } from "../../../../containers/IndexForm";
import { EuiButton, EuiButtonEmpty, EuiCallOut, EuiFlexGroup, EuiFlexItem, EuiSpacer, EuiTitle, EuiLink } from "@elastic/eui";
import CustomFormRow from "../../../../components/CustomFormRow";
import { ContentPanel } from "../../../../components/ContentPanel";
import FormGenerator, { AllBuiltInComponents, IFormGeneratorRef } from "../../../../components/FormGenerator";
import { Alias } from "../../../../../server/models/interfaces";
import useField from "../../../../lib/field";
import { getIndexDetail, getOptions, getRolloveredIndex, onSubmit, submitWriteIndex } from "../../hooks";
import { IRolloverRequestBody } from "../../interface";
import { filterByMinimatch } from "../../../../../utils/helper";
import { SYSTEM_ALIAS } from "../../../../../utils/constants";

export interface RolloverProps extends RouteComponentProps<{ source?: string }> {}

export default function Rollover(props: RolloverProps) {
  const [options, setOptions] = useState<{
    alias: { label: string; aliases: Alias[] }[];
    dataStreams: { label: string }[];
  }>({
    alias: [],
    dataStreams: [],
  });
  const coreService = useContext(CoreServicesContext);
  const services = useContext(ServicesContext) as BrowserServices;
  const sourceRef = useRef<IFormGeneratorRef>(null);
  const indexFormRef = useRef<IndexForm>(null);
  const [tempValue, setValue] = useState<IRolloverRequestBody>({
    source: props.match.params.source,
  });
  const [loading, setIsLoading] = useState(false);
  const [writeIndexValue, setWriteIndexValue] = useState<string>("");
  const field = useField({
    values: {
      targetIndex: {
        settings: {
          "index.number_of_shards": 1,
          "index.number_of_replicas": 1,
          "index.refresh_interval": "1s",
        },
      },
    },
  });

  const onChange = (val?: Record<string, any>) => {
    const finalResult = merge({}, tempValue, val);
    setValue(finalResult);
  };

  const refreshOptions = async () => {
    const result = await getOptions({
      services,
    });
    if (result.ok) {
      setOptions(
        result.response || {
          alias: [],
          dataStreams: [],
        }
      );
    } else {
      coreService?.notifications.toasts.addDanger(result.error || "");
    }
  };

  useEffect(() => {
    coreService?.chrome.setBreadcrumbs([
      BREADCRUMBS.INDEX_MANAGEMENT,
      {
        ...BREADCRUMBS.ROLLOVER,
        href: `#${props.location.pathname}`,
      },
    ]);
    refreshOptions();
  }, []);

  const sourceOptions = useMemo(() => {
    return [
      {
        label: "aliases",
        options: options.alias,
      },
      {
        label: "data streams",
        options: options.dataStreams,
      },
    ];
  }, [options]);

  const initialValue = useMemo(() => {
    return {
      source: props.match.params.source,
    };
  }, [props.match.params.source]);

  const sourceType: "dataStreams" | "alias" | undefined = useMemo(() => {
    const sourceValue = sourceRef.current?.getValue("source");
    if (options.alias.find((item) => item.label === sourceValue)) {
      return "alias";
    }

    if (options.dataStreams.find((item) => item.label === sourceValue)) {
      return "dataStreams";
    }

    return;
  }, [sourceRef.current?.getValue("source")]);

  const writingIndex = (() => {
    const findItem = options.alias.find((item) => item.label === sourceRef.current?.getValue("source"));
    let writeIndex = "";
    if (findItem) {
      if (findItem.aliases.length > 1) {
        // has to check if it has write_index
        if (findItem.aliases.some((item) => item.is_write_index === "true")) {
          const indexItem = findItem.aliases.find((item) => item.is_write_index === "true");
          writeIndex = indexItem?.index || "";
        }
      } else {
        writeIndex = findItem.aliases[0].index;
      }
    }

    return writeIndex;
  })();

  useEffect(() => {
    if (writingIndex && tempValue.source) {
      // do a dry run to get the rollovered writing index
      getRolloveredIndex({
        alias: tempValue.source,
        services,
      }).then((result) => {
        if (result && result.ok) {
          field.setValue(["targetIndex", "index"], result.response);
        } else {
          field.setValue(["targetIndex", "index"], "");
        }
      });
      indexFormRef.current?.simulateFromTemplate();
    }
  }, [writingIndex]);

  const reasons = useMemo(() => {
    let result: React.ReactChild[] = [];
    if (sourceType === "alias") {
      const findItem = options.alias.find((item) => item.label === sourceRef.current?.getValue("source"));
      if (!writingIndex) {
        result.push(
          <>
            <EuiFlexGroup alignItems="flexEnd">
              <EuiFlexItem grow={false}>
                Select a write index from this alias before performing rollover.
                <EuiSpacer size="s" />
                <CustomFormRow label="Select an index from this alias">
                  <AllBuiltInComponents.ComboBoxSingle
                    placeholder="Select an index"
                    value={writeIndexValue ? writeIndexValue : undefined}
                    onChange={(val) => setWriteIndexValue(val)}
                    options={findItem?.aliases?.map((item) => ({ label: item.index }))}
                  />
                </CustomFormRow>
              </EuiFlexItem>
              <EuiFlexItem grow={false}>
                <EuiButton
                  disabled={!writeIndexValue}
                  fill
                  color="primary"
                  onClick={async () => {
                    const result = await submitWriteIndex({
                      services,
                      writeIndexValue: writeIndexValue,
                      sourceRef,
                    });
                    if (result.ok) {
                      coreService?.notifications.toasts.addSuccess(`Set ${writeIndexValue} as write index successfully.`);
                      refreshOptions();
                    } else {
                      coreService?.notifications.toasts.addDanger(result.error);
                    }
                  }}
                >
                  Assign as write index
                </EuiButton>
              </EuiFlexItem>
            </EuiFlexGroup>
          </>
        );
      }
    }

    return result;
  }, [sourceType, options, writeIndexValue]);

  return (
    <div style={{ padding: "0 50px" }}>
      <EuiTitle>
        <h1>Roll over</h1>
      </EuiTitle>
      <CustomFormRow
        helpText="Rollover creates a new writing index for a data stream or index alias."
        style={{
          marginBottom: 20,
        }}
      >
        <></>
      </CustomFormRow>
      <ContentPanel title="Configure source" titleSize="s">
        {sourceType === "alias" && filterByMinimatch(tempValue.source || "", SYSTEM_ALIAS) ? (
          <>
            <EuiCallOut color="warning">
              This alias may contain critical system data. Rollovering system aliases may break OpenSearch.
            </EuiCallOut>
            <EuiSpacer />
          </>
        ) : null}
        <FormGenerator
          value={initialValue}
          onChange={onChange}
          formFields={[
            {
              name: "source",
              rowProps: {
                label: "Select an alias or data stream",
                helpText: "Select an alias or data stream to roll over.",
              },
              options: {
                rules: [
                  {
                    required: true,
                    message: "One alias or data stream should be specified.",
                  },
                ],
                props: {
                  options: sourceOptions,
                  placeholder: "Select a alias or data stream.",
                },
              },
              type: "ComboBoxSingle",
            },
          ]}
          ref={sourceRef}
        />
        {reasons.length ? (
          <>
            <EuiSpacer />
            <EuiCallOut color="danger" title="This alias does not contain a write index.">
              {reasons.map((item, index) => (
                <div key={index}>{item}</div>
              ))}
            </EuiCallOut>
          </>
        ) : null}
        {(() => {
          if (sourceType === "alias" && writingIndex) {
            return (
              <>
                <EuiSpacer />
                <CustomFormRow label="Assigned write index">
                  <span>{writingIndex}</span>
                </CustomFormRow>
              </>
            );
          }

          return null;
        })()}
      </ContentPanel>
      <EuiSpacer />
      {sourceType === "alias" && writingIndex ? (
        <>
          <ContentPanel
            title="Configure new rollover index"
            titleSize="s"
            actions={
              <EuiButton
                onClick={() => {
                  indexFormRef.current?.importSettings({
                    index: writingIndex,
                  });
                }}
              >
                Import from old write index
              </EuiButton>
            }
          >
            <IndexFormWrapper
              {...field.registerField({
                name: "targetIndex",
              })}
              onGetIndexDetail={async (indexName) => {
                try {
                  return await getIndexDetail({
                    services,
                    indexName,
                    newIndexName: indexFormRef.current?.getValue()?.index || "",
                  });
                } catch (e) {
                  coreService?.notifications.toasts.addDanger(e as string);
                  throw e;
                }
              }}
              ref={indexFormRef}
              withoutPanel
            />
          </ContentPanel>
          <EuiSpacer />
        </>
      ) : null}
      <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            onClick={() => {
              if (sourceType === "alias") {
                props.history.push(ROUTES.ALIASES);
              } else {
                props.history.push(ROUTES.DATA_STREAMS);
              }
            }}
            data-test-subj="rolloverCancelButton"
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            disabled={loading || (sourceType === "alias" && !writingIndex)}
            fill
            onClick={async () => {
              setIsLoading(true);
              const result = await onSubmit({
                services,
                sourceRef,
                sourceType,
                writingIndex,
                indexFormRef,
                tempValue,
              });
              setIsLoading(false);
              if (result.ok) {
                coreService?.notifications.toasts.addSuccess({
                  title: ((
                    <>
                      <div>{tempValue.source} has been successfully rolled over.</div>
                      {result.response?.new_index ? (
                        <div>
                          <EuiLink href={`#${ROUTES.INDEX_DETAIL}/${result.response.new_index}`}>{result.response?.new_index}</EuiLink> is
                          now the latest write index.
                        </div>
                      ) : null}
                    </>
                  ) as unknown) as string,
                });
                props.history.replace(sourceType === "alias" ? ROUTES.ALIASES : ROUTES.DATA_STREAMS);
              } else {
                coreService?.notifications.toasts.addDanger(result.error);
              }
            }}
            isLoading={loading}
            data-test-subj="rolloverSubmitButton"
          >
            Roll over
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
