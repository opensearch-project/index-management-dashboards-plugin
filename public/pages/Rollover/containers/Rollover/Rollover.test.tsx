/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, waitFor } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { Route, RouteComponentProps, Switch } from "react-router-dom";
import { browserServicesMock, coreServicesMock, apiCallerMock } from "../../../../../test/mocks";
import Rollover, { RolloverProps } from "./Rollover";
import { ModalProvider } from "../../../../components/Modal";
import { ServicesContext } from "../../../../services";
import { CoreServicesContext } from "../../../../components/core_services";

function renderWithRouter(props: Omit<RolloverProps, keyof RouteComponentProps>, initialEntries: string[]) {
  return {
    ...render(
      <Router initialEntries={initialEntries}>
        <CoreServicesContext.Provider value={coreServicesMock}>
          <ServicesContext.Provider value={browserServicesMock}>
            <ModalProvider>
              <Switch>
                <Route path="/:source" render={(routeProps) => <Rollover {...props} {...routeProps} />} />
              </Switch>
            </ModalProvider>
          </ServicesContext.Provider>
        </CoreServicesContext.Provider>
      </Router>
    ),
  };
}

describe("container <Rollover /> spec", () => {
  beforeEach(() => {
    apiCallerMock(browserServicesMock);
  });
  it("render the component", async () => {
    browserServicesMock.indexService.getDataStreams = jest.fn(() => {
      return {
        ok: true,
      } as any;
    }) as typeof browserServicesMock.indexService.getDataStreams;
    const { container } = renderWithRouter({}, [`/test_alias`]);

    await waitFor(() => {
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
