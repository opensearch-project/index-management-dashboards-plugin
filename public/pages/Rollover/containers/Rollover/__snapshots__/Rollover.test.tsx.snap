// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container <Rollover /> spec render the component 1`] = `
<div
  style="padding: 0px 50px;"
>
  <h1
    class="euiTitle euiTitle--medium"
  >
    Roll over
  </h1>
  <div
    class="euiFormRow"
    id="some_html_id-row"
    style="margin-bottom: 20px;"
  >
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFormHelpText euiFormRow__text"
        style="padding-top: 0px; padding-bottom: 4px;"
      >
        Rollover creates a new writing index for a data stream or index alias.
      </div>
    </div>
  </div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Configure source
          <span
            class="panel-header-count"
          >
             
             
          </span>
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        class="euiForm"
      >
        <div
          class="euiFormRow"
          data-test-subj="form-name-source"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              Select an alias or data stream
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormHelpText euiFormRow__text"
              style="padding-top: 0px; padding-bottom: 4px;"
            >
              Select an alias or data stream to roll over.
            </div>
            <span
              class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                class="euiComboBox"
                role="combobox"
              >
                <div
                  class="euiFormControlLayout"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <div
                      class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                      data-test-subj="comboBoxInput"
                      tabindex="-1"
                    >
                      <span
                        class="euiComboBoxPill euiComboBoxPill--plainText"
                      >
                        test_alias
                      </span>
                      <div
                        class="euiComboBox__input"
                        style="font-size: 14px; display: inline-block;"
                      >
                        <input
                          aria-controls=""
                          data-test-subj="comboBoxSearchInput"
                          role="textbox"
                          style="box-sizing: content-box; width: 2px;"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                    <div
                      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                    >
                      <button
                        aria-label="Clear input"
                        class="euiFormControlLayoutClearButton"
                        data-test-subj="comboBoxClearButton"
                        type="button"
                      >
                        EuiIconMock
                      </button>
                      <button
                        aria-label="Open list of options"
                        class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                        data-test-subj="comboBoxToggleListButton"
                        type="button"
                      >
                        EuiIconMock
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButtonEmpty euiButtonEmpty--primary"
        data-test-subj="rolloverCancelButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            Cancel
          </span>
        </span>
      </button>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButton euiButton--primary euiButton--fill"
        data-test-subj="rolloverSubmitButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Roll over
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
